src/rpc.rs

```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
     }
 }
 
+enum DisconnectorState
+{
+    Connected,
+    Disconnecting,
+    Disconnected
+}
+
+pub struct Disconnector<VatId> where VatId: 'static {
+    connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>,
+    state:  DisconnectorState,
+}
+
+impl <VatId> Disconnector<VatId> {
+    pub fn new(connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>) -> Disconnector<VatId> {
+        let state = match *(connection_state.borrow()) {
+            Some(_) => {
+                DisconnectorState::Connected
+            },
+            None => {
+                DisconnectorState::Disconnected
+            }
+        };
+        Disconnector {
+            connection_state: connection_state,
+            state: state,
+        }
+    }
+    fn disconnect(&self) {
+        if let Some(ref state) = *(self.connection_state.borrow()) {
+            state.disconnect(::capnp::Error::disconnected("client requested disconnect".to_owned()));
+        }
+    }
+}
+
+impl <VatId> Future for Disconnector<VatId>
+where
+    VatId: 'static,
+{
+    type Item = ();
+    type Error = ::capnp::Error;
+
+    fn poll(&mut self) -> ::futures::Poll<Self::Item, Self::Error> {
+        self.state = match self.state {
+            DisconnectorState::Connected => {
+                self.disconnect();
+                DisconnectorState::Disconnecting
+            },
+            DisconnectorState::Disconnecting => {
+                if let Some(_) = *(self.connection_state.borrow()) {
+                    DisconnectorState::Disconnecting
+                } else {
+                    DisconnectorState::Disconnected
+                }
+            },
+            DisconnectorState::Disconnected => {
+                DisconnectorState::Disconnected
+            }
+        };
+        match self.state {
+            DisconnectorState::Connected => unreachable!(),
+            DisconnectorState::Disconnecting => {
+                ::futures::task::current().notify();
```

Comment:
> Like, an Rc<Option<Task>> (shared with Disconnector) that was along side the RpcSystem::connection_state that you'd call notify on when you set the connection_state to None?

Yeah, something like that is what I had in mind. I suppose it would actually need to be a `Rc<Vec<Task>>`, in case there are multiple `Disconnector`s.

Your call on whether you want to bother with this or not.